File: enode/src/main/java/com/enode/infrastructure/impl/DefaultMessageHandlerProvider.java
package com.enode.infrastructure.impl;

import com.enode.common.container.IObjectContainer;
import com.enode.infrastructure.IMessage;
import com.enode.infrastructure.IMessageHandler;
import com.enode.infrastructure.IMessageHandlerProvider;
import com.enode.infrastructure.IMessageHandlerProxy1;

import javax.inject.Inject;
import java.lang.reflect.Method;

public class DefaultMessageHandlerProvider extends AbstractHandlerProvider<Class, IMessageHandlerProxy1, Class> implements IMessageHandlerProvider {
    private IObjectContainer objectContainer;

    @Inject
    public DefaultMessageHandlerProvider(IObjectContainer objectContainer) {
        this.objectContainer = objectContainer;
    }

    @Override
    protected Class getGenericHandlerType() {
        return IMessageHandler.class;
    }

    @Override
    protected Class getKey(Method method) {
        return method.getParameterTypes()[0];
    }

    @Override
    protected Class<? extends IMessageHandlerProxy1> getHandlerProxyImplementationType() {
        return MessageHandlerProxy1.class;
    }

    @Override
    protected boolean isHandlerSourceMatchKey(Class handlerSource, Class key) {
        return key.isAssignableFrom(handlerSource);
    }

    @Override
    protected boolean isHandleMethodMatch(Method method) {
        if (!"handleAsync".equals(method.getName())) {
            return false;
        }
        if (method.getParameterTypes().length != 1) {
            return false;
        }
        if (IMessage.class.equals(method.getParameterTypes()[0])) {
            return false;
        }
        if (!IMessage.class.isAssignableFrom(method.getParameterTypes()[0])) {
            return false;
        }
        return true;
    }

    @Override
    protected IObjectContainer getObjectContainer() {
        return objectContainer;
    }
}

File: enode/src/main/java/org/enodeframework/messaging/impl/DefaultMessageHandlerProvider.kt
package org.enodeframework.messaging.impl

import org.enodeframework.common.container.IObjectContainer
import org.enodeframework.common.container.ObjectContainer
import org.enodeframework.infrastructure.impl.AbstractHandlerProvider
import org.enodeframework.messaging.IMessage
import org.enodeframework.messaging.IMessageHandlerProvider
import org.enodeframework.messaging.IMessageHandlerProxy1
import java.lang.reflect.Method
import kotlin.reflect.jvm.kotlinFunction

/**
 * @author anruence@gmail.com
 */
class DefaultMessageHandlerProvider : AbstractHandlerProvider<Class<*>?, IMessageHandlerProxy1?, Class<*>?>(), IMessageHandlerProvider {
    override fun getKey(method: Method): Class<*> {
        return method.parameterTypes[0]
    }

    override fun getHandlerProxyImplementationType(): Class<out IMessageHandlerProxy1> {
        return MessageHandlerProxy1::class.java
    }

    override fun isHandleMethodMatch(method: Method): Boolean {
        val paramCount = method.parameterTypes.size
        if (paramCount != 1) {
            if (!isSuspendMethod(method)) {
                return false
            }
        }
        if (IMessage::class.java == method.parameterTypes[0]) {
            return false
        }
        return if (!IMessage::class.java.isAssignableFrom(method.parameterTypes[0])) {
            false
        } else isMethodAnnotationSubscribe(method)
    }

    override fun isSuspendMethod(method: Method): Boolean {
        return method.kotlinFunction?.isSuspend == true
    }

    override fun getObjectContainer(): IObjectContainer {
        return ObjectContainer.INSTANCE
    }

    override fun isHandlerSourceMatchKey(handlerSource: Class<*>?, key: Class<*>?): Boolean {
        return key == handlerSource
    }
}